### PR TITLE
feat: open ai example

### PR DESCRIPTION
**Pull request by thantos**
Simple TODO Open AI chat plugin example which follows: https://platform.openai.com/docs/plugins/examples

* `/.well-known/ai-plugin.json` - open ai plugin config
* `/.well-known/openapi.json` - openapi api schema
* GET /:username/todos
* POST /:username/todos
* DELETE /:username/todos/{todoId}
* TODO - test with langchain
* TOOD - test with chatgpt

### CODE REVIEW

**Review by sam-goodwin — CHANGES_REQUESTED** [review]
They're called chat gpt plug-ins, not open so plug-ins.

File: examples/open-ai-plugin/packages/service/src/todo.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import { extendApi } from "@anatine/zod-openapi";
+import { command, entity } from "@eventual/core";
+import { ulid } from "ulidx";
+import z from "zod";
+
+const userNameSchema = extendApi(z.string(), {
```
Comment (sam-goodwin) [inline examples/open-ai-plugin/packages/service/src/todo.ts:6]: I really don't like the extend api ergonomics. Module augmentation of zod would be better imo.

**Review by sam-goodwin — CHANGES_REQUESTED** [review]

File: examples/open-ai-plugin/packages/service/src/todo.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    params: {
+      username: "path",
+    },
+    input: todoSchema.omit({ todoId: true }),
+    output: todoSchema,
+  },
+  async (request) => {
+    const id = ulid();
+
+    const todo: z.infer<typeof todoSchema> = {
```
Comment (sam-goodwin) [inline examples/open-ai-plugin/packages/service/src/todo.ts:41]: Poor use of infer in example. Extract out to an interface or type next to the schema.

**Review by sam-goodwin — CHANGES_REQUESTED** [review]

File: examples/open-ai-plugin/packages/service/src/todo.ts (hunk 1 above)
Comment (thantos) [reply examples/open-ai-plugin/packages/service/src/todo.ts:6]: `@anatine/zod-openapi` only supports this method.

There was another zod to opanapi library that supported module augmentation, but the inclusion of a side effect statement would be more complex than a single import and wrapping the types.

Comment (thantos) [reply examples/open-ai-plugin/packages/service/src/todo.ts:6]: https://github.com/anatine/zod-plugins/issues/82

**Comment by thantos** [comment]
> They're called chat gpt plug-ins, not open so plug-ins.

Updated to `open-api-chat-plugin`

File: examples/open-ai-chat-plugin/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+This is where you control what infrastructure is deployed with your Service, for example adding DynamoDB Tables, SQS Queues, or other stateful Resources.
+
+### `packages/service`
+
+This is where you add business logic such as APIs, Event handlers, Workflows and Tasks.
+
+## Deployed Infrastructure
+
+After deploying to AWS, you'll have a single stack named `open-ai-plugin` containing your Service. Take a look at the structure using the Resources view in CloudFormation. Here, you can find a list of all the Lambda Functions and other Resources that come with a Service.
```
Comment (sam-goodwin) [inline examples/open-ai-chat-plugin/README.md:23]: Can we call it a chat gpt plug-in? It's not an open ai plug-in.

**Review by sam-goodwin — CHANGES_REQUESTED** [review]

**Comment by sam-goodwin** [comment]
> > They're called chat gpt plug-ins, not open so plug-ins.
> 
> 
> 
> Updated to `open-api-chat-plugin`

Lol I think you mean open-ai 😅. Happens to me all the time

File: examples/open-ai-chat-plugin/README.md (hunk 3 above)
Comment (thantos) [reply examples/open-ai-chat-plugin/README.md:23]: Not according to their docs. 

They call them chat plugins or openai plugins 

> OpenAI plugins connect ChatGPT to third-party applications.

https://platform.openai.com/docs/plugins/introduction

Comment (thantos) [reply examples/open-ai-chat-plugin/README.md:23]: Looks like they have not decided what to call them. Ha... The format is usable for multiple purposes, chat is just one of them. This example could be called chatgpt plugin though it's not quite chat gpt when used with lang chain.

File: examples/open-ai-chat-plugin/packages/service/src/todo.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import { ulid } from "ulidx";
+import z from "zod";
+
+const userNameSchema = extendApi(z.string(), {
+  description: "The name of the user",
+});
+
+const todoIdSchema = extendApi(z.string(), {
+  description: "Id of a single todo item",
+});
```
Comment (sam-goodwin) [inline examples/open-ai-chat-plugin/packages/service/src/todo.ts:12]: This is so ugly. Why can't we use the built in zod description?

```ts
z.string().description("ID of a single TODO item")
```

File: examples/open-ai-chat-plugin/packages/service/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+export interface AIPluginAuth {
+  type: string | "none";
+  authorization_type?: string;
+  authorization_url?: string;
+  client_url?: string;
+  scope?: string;
+  authorization_content_type?: string;
+  verification_tokens?: AIPluginVerificationTokens;
+  instructions?: string;
+}
```
Comment (sam-goodwin) [inline examples/open-ai-chat-plugin/packages/service/src/types.ts:33]: Is there an existing type library we can use? If not, should we publish one?

**Review by sam-goodwin — APPROVED** [review]

File: examples/open-ai-chat-plugin/packages/service/src/todo.ts (hunk 4 above)
Comment (thantos) [reply examples/open-ai-chat-plugin/packages/service/src/todo.ts:12]: It doesn't work unfortunately.

**Pull request merged**